File: server_python/routes/interviews.py
"""
Interviews routes for managing interview sessions and analysis
"""

import uuid
import json
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, HTTPException, Depends, Query, status
from slowapi import Limiter
from slowapi.util import get_remote_address
from config.database import database, interviews_table, candidates_table, job_positions_table
from models.schemas import InterviewCreate, InterviewUpdate, InterviewResponse, MessageResponse
from middleware.auth import get_current_active_user, require_roles
from services.groq_service import groq_service
from sqlalchemy import select, update, delete, func, or_

router = APIRouter()
limiter = Limiter(key_func=get_remote_address)

@router.post("/", response_model=dict, status_code=status.HTTP_201_CREATED)
@limiter.limit("20/minute")
async def create_interview(
    request,
    interview_data: InterviewCreate,
    current_user: dict = Depends(require_roles(["admin", "recruiter", "interviewer"]))
):
    """Create a new interview session"""
    
    # Verify candidate exists
    candidate_query = select(candidates_table).where(candidates_table.c.id == interview_data.candidate_id)
    candidate = await database.fetch_one(candidate_query)
    
    if not candidate:
        raise HTTPException(
            status_code=404,
            detail="Candidate not found"
        )
    
    # Verify job position exists
    job_query = select(job_positions_table).where(job_positions_table.c.id == interview_data.job_position_id)
    job = await database.fetch_one(job_query)
    
    if not job:
        raise HTTPException(
            status_code=404,
            detail="Job position not found"
        )
    
    interview_id = str(uuid.uuid4())
    
    insert_query = interviews_table.insert().values(
        id=interview_id,
        candidate_id=interview_data.candidate_id,
        job_position_id=interview_data.job_position_id,
        interviewer_id=current_user["id"],
        scheduled_at=interview_data.scheduled_at,
        status="scheduled"
    )
    
    try:
        await database.execute(insert_query)
        
        # Get created interview with related data
        interview_query = """
        SELECT i.*, c.name as candidate_name, j.title as job_title, u.name as interviewer_name
        FROM interviews i
        JOIN candidates c ON i.candidate_id = c.id
        JOIN job_positions j ON i.job_position_id = j.id
        JOIN users u ON i.interviewer_id = u.id
        WHERE i.id = ?
        """
        
        created_interview = await database.fetch_one(interview_query, [interview_id])
        
        return {
            "message": "Interview created successfully",
            "interview": dict(created_interview)
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to create interview: {str(e)}"
        )

@router.get("/", response_model=dict)
async def get_interviews(
    page: int = Query(1, ge=1),
    limit: int = Query(10, ge=1, le=100),
    status_filter: Optional[str] = Query(None, regex="^(scheduled|in_progress|completed|cancelled)$"),
    search: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_active_user)
):
    """Get all interviews with pagination and filtering"""
    
    offset = (page - 1) * limit
    
    # Base query with joins
    base_query = """
    FROM interviews i
    JOIN candidates c ON i.candidate_id = c.id
    JOIN job_positions j ON i.job_position_id = j.id
    JOIN users u ON i.interviewer_id = u.id
    """
    
    where_conditions = []
    params = []
    
    # Filter by status
    if status_filter:
        where_conditions.append("i.status = ?")
        params.append(status_filter)
    
    # Search filter
    if search:
        where_conditions.append("(c.name ILIKE ? OR j.title ILIKE ? OR u.name ILIKE ?)")
        params.extend([f"%{search}%", f"%{search}%", f"%{search}%"])
    
    # Role-based filtering
    if current_user["role"] == "interviewer":
        where_conditions.append("i.interviewer_id = ?")
        params.append(current_user["id"])
    
    where_clause = " WHERE " + " AND ".join(where_conditions) if where_conditions else ""
    
    # Count query
    count_query = f"SELECT COUNT(*) {base_query} {where_clause}"
    total = await database.fetch_val(count_query, params)
    
    # Main query
    main_query = f"""
    SELECT i.*, c.name as candidate_name, c.email as candidate_email,
           j.title as job_title, u.name as interviewer_name
    {base_query}
    {where_clause}
    ORDER BY i.created_at DESC
    LIMIT ? OFFSET ?
    """
    
    params.extend([limit, offset])
    interviews = await database.fetch_all(main_query, params)
    
    # Process interviews data
    processed_interviews = []
    for interview in interviews:
        interview_dict = dict(interview)
        
        # Parse JSON fields
        if interview_dict.get("questions"):
            try:
                interview_dict["questions"] = json.loads(interview_dict["questions"])
            except (json.JSONDecodeError, TypeError):
                interview_dict["questions"] = None
        
        if interview_dict.get("answers"):
            try:
                interview_dict["answers"] = json.loads(interview_dict["answers"])
            except (json.JSONDecodeError, TypeError):
                interview_dict["answers"] = None
        
        if interview_dict.get("ai_analysis"):
            try:
                interview_dict["ai_analysis"] = json.loads(interview_dict["ai_analysis"])
            except (json.JSONDecodeError, TypeError):
                interview_dict["ai_analysis"] = None
        
        processed_interviews.append(interview_dict)
    
    return {
        "interviews": processed_interviews,
        "pagination": {
            "page": page,
            "limit": limit,
            "total": total,
            "pages": (total + limit - 1) // limit
        }
    }

@router.get("/{interview_id}", response_model=dict)
async def get_interview(
    interview_id: str,
    current_user: dict = Depends(get_current_active_user)
):
    """Get interview by ID"""
    
    query = """
    SELECT i.*, c.name as candidate_name, c.email as candidate_email, c.phone as candidate_phone,
           c.skills as candidate_skills, c.experience_years as candidate_experience,
           j.title as job_title, j.description as job_description,
           u.name as interviewer_name
    FROM interviews i
    JOIN candidates c ON i.candidate_id = c.id
    JOIN job_positions j ON i.job_position_id = j.id
    JOIN users u ON i.interviewer_id = u.id
    WHERE i.id = ?
    """
    
    interview = await database.fetch_one(query, [interview_id])
    
    if not interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to view this interview"
        )
    
    interview_dict = dict(interview)
    
    # Parse JSON fields
    if interview_dict.get("questions"):
        try:
            interview_dict["questions"] = json.loads(interview_dict["questions"])
        except (json.JSONDecodeError, TypeError):
            interview_dict["questions"] = None
    
    if interview_dict.get("answers"):
        try:
            interview_dict["answers"] = json.loads(interview_dict["answers"])
        except (json.JSONDecodeError, TypeError):
            interview_dict["answers"] = None
    
    if interview_dict.get("ai_analysis"):
        try:
            interview_dict["ai_analysis"] = json.loads(interview_dict["ai_analysis"])
        except (json.JSONDecodeError, TypeError):
            interview_dict["ai_analysis"] = None
    
    if interview_dict.get("candidate_skills"):
        try:
            interview_dict["candidate_skills"] = json.loads(interview_dict["candidate_skills"])
        except (json.JSONDecodeError, TypeError):
            interview_dict["candidate_skills"] = []
    
    return {"interview": interview_dict}

@router.put("/{interview_id}", response_model=MessageResponse)
async def update_interview(
    interview_id: str,
    interview_data: InterviewUpdate,
    current_user: dict = Depends(get_current_active_user)
):
    """Update interview information"""
    
    # Check if interview exists and user has permission
    query = select(interviews_table).where(interviews_table.c.id == interview_id)
    existing_interview = await database.fetch_one(query)
    
    if not existing_interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        existing_interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to update this interview"
        )
    
    # Prepare update data
    update_data = {}
    if interview_data.questions is not None:
        update_data["questions"] = interview_data.questions
    if interview_data.answers is not None:
        update_data["answers"] = interview_data.answers
    if interview_data.ai_analysis is not None:
        update_data["ai_analysis"] = interview_data.ai_analysis
    if interview_data.score is not None:
        update_data["score"] = interview_data.score
    if interview_data.status is not None:
        update_data["status"] = interview_data.status.value
        # Set completed_at if status is completed
        if interview_data.status.value == "completed":
            update_data["completed_at"] = datetime.utcnow()
    if interview_data.scheduled_at is not None:
        update_data["scheduled_at"] = interview_data.scheduled_at
    if interview_data.completed_at is not None:
        update_data["completed_at"] = interview_data.completed_at
    
    if not update_data:
        raise HTTPException(
            status_code=400,
            detail="No valid fields provided for update"
        )
    
    # Update interview
    update_query = (
        update(interviews_table)
        .where(interviews_table.c.id == interview_id)
        .values(**update_data)
    )
    
    await database.execute(update_query)
    
    return MessageResponse(message="Interview updated successfully")

@router.delete("/{interview_id}", response_model=MessageResponse)
async def delete_interview(
    interview_id: str,
    current_user: dict = Depends(require_roles(["admin", "recruiter"]))
):
    """Delete interview"""
    
    # Check if interview exists
    query = select(interviews_table).where(interviews_table.c.id == interview_id)
    existing_interview = await database.fetch_one(query)
    
    if not existing_interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Delete interview
    delete_query = delete(interviews_table).where(interviews_table.c.id == interview_id)
    await database.execute(delete_query)
    
    return MessageResponse(message="Interview deleted successfully")

@router.post("/{interview_id}/generate-questions", response_model=dict)
async def generate_questions_for_interview(
    interview_id: str,
    difficulty: str = Query("medium", regex="^(easy|medium|hard)$"),
    current_user: dict = Depends(get_current_active_user)
):
    """Generate interview questions for a specific interview"""
    
    # Get interview with related data
    query = """
    SELECT i.*, c.skills as candidate_skills, j.title as job_title, j.description as job_description
    FROM interviews i
    JOIN candidates c ON i.candidate_id = c.id
    JOIN job_positions j ON i.job_position_id = j.id
    WHERE i.id = ?
    """
    
    interview = await database.fetch_one(query, [interview_id])
    
    if not interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to generate questions for this interview"
        )
    
    try:
        # Parse candidate skills
        candidate_skills = []
        if interview["candidate_skills"]:
            try:
                candidate_skills = json.loads(interview["candidate_skills"])
            except (json.JSONDecodeError, TypeError):
                candidate_skills = []
        
        # Generate interview questions using AI
        questions = await groq_service.generate_interview_questions(
            job_title=interview["job_title"],
            job_description=interview["job_description"] or "",
            candidate_skills=candidate_skills,
            difficulty=difficulty
        )
        
        # Update interview with generated questions
        update_query = (
            update(interviews_table)
            .where(interviews_table.c.id == interview_id)
            .values(questions=json.dumps(questions.dict()))
        )
        
        await database.execute(update_query)
        
        return {
            "message": "Interview questions generated successfully",
            "interview_id": interview_id,
            "job_title": interview["job_title"],
            "difficulty": difficulty,
            "questions": questions.dict()
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to generate interview questions: {str(e)}"
        )

@router.post("/{interview_id}/analyze-response", response_model=dict)
async def analyze_interview_response(
    interview_id: str,
    question: str,
    answer: str,
    current_user: dict = Depends(get_current_active_user)
):
    """Analyze a specific interview response using AI"""
    
    # Get interview with job context
    query = """
    SELECT i.*, j.title as job_title, j.description as job_description
    FROM interviews i
    JOIN job_positions j ON i.job_position_id = j.id
    WHERE i.id = ?
    """
    
    interview = await database.fetch_one(query, [interview_id])
    
    if not interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to analyze responses for this interview"
        )
    
    try:
        # Prepare job context
        job_context = f"Job Title: {interview['job_title']}\nJob Description: {interview['job_description'] or 'Not provided'}"
        
        # Analyze response using AI
        analysis = await groq_service.analyze_interview_response(
            question=question,
            answer=answer,
            job_context=job_context
        )
        
        return {
            "message": "Interview response analyzed successfully",
            "interview_id": interview_id,
            "question": question,
            "answer": answer,
            "analysis": analysis.dict()
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to analyze interview response: {str(e)}"
        )

@router.post("/{interview_id}/start", response_model=MessageResponse)
async def start_interview(
    interview_id: str,
    current_user: dict = Depends(get_current_active_user)
):
    """Start an interview session"""
    
    # Check if interview exists and user has permission
    query = select(interviews_table).where(interviews_table.c.id == interview_id)
    interview = await database.fetch_one(query)
    
    if not interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to start this interview"
        )
    
    if interview["status"] != "scheduled":
        raise HTTPException(
            status_code=400,
            detail=f"Interview cannot be started. Current status: {interview['status']}"
        )
    
    # Update interview status
    update_query = (
        update(interviews_table)
        .where(interviews_table.c.id == interview_id)
        .values(status="in_progress")
    )
    
    await database.execute(update_query)
    
    return MessageResponse(message="Interview started successfully")

@router.post("/{interview_id}/complete", response_model=MessageResponse)
async def complete_interview(
    interview_id: str,
    final_score: Optional[int] = Query(None, ge=0, le=100),
    current_user: dict = Depends(get_current_active_user)
):
    """Complete an interview session"""
    
    # Check if interview exists and user has permission
    query = select(interviews_table).where(interviews_table.c.id == interview_id)
    interview = await database.fetch_one(query)
    
    if not interview:
        raise HTTPException(
            status_code=404,
            detail="Interview not found"
        )
    
    # Role-based access control
    if (current_user["role"] == "interviewer" and 
        interview["interviewer_id"] != current_user["id"]):
        raise HTTPException(
            status_code=403,
            detail="You don't have permission to complete this interview"
        )
    
    if interview["status"] not in ["scheduled", "in_progress"]:
        raise HTTPException(
            status_code=400,
            detail=f"Interview cannot be completed. Current status: {interview['status']}"
        )
    
    # Prepare update data
    update_data = {
        "status": "completed",
        "completed_at": datetime.utcnow()
    }
    
    if final_score is not None:
        update_data["score"] = final_score
    
    # Update interview
    update_query = (
        update(interviews_table)
        .where(interviews_table.c.id == interview_id)
        .values(**update_data)
    )
    
    await database.execute(update_query)
    
    return MessageResponse(message="Interview completed successfully")

File: server/routes/interviews.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const Joi = require('joi');
const database = require('../config/database');
const { authenticateToken } = require('../middleware/auth');
const groqService = require('../services/groqService');

const router = express.Router();
const db = database.getDb();

// Validation schemas
const interviewSchema = Joi.object({
  candidateId: Joi.string().required(),
  jobPositionId: Joi.string().required(),
  scheduledAt: Joi.date().iso().required(),
  questions: Joi.array().items(Joi.object()).optional()
});

const answerSchema = Joi.object({
  questionId: Joi.string().required(),
  question: Joi.string().required(),
  answer: Joi.string().required()
});

router.delete('/:interviewId', authenticateToken, async (req, res) => {
  const interviewId = req.params.interviewId;

  db.run('DELETE FROM interviews WHERE id = ?', [interviewId], function (err) {
    if (err) {
      console.error('Error deleting interviewId:', err.message);
      return res.status(500).json({ success: false, error: 'Failed to delete scheduled interview' });
    }

    // this.changes tells you how many rows were deleted
    if (this.changes === 0) {
      return res.status(404).json({ success: false, error: 'interviewId not found' });
    }

    res.json({ success: true });
  });
});

// Schedule interview
router.post('/', authenticateToken, async (req, res) => {
  try {
    const { error, value } = interviewSchema.validate(req.body);
    if (error) {
      return res.status(400).json({ error: error.details[0].message });
    }

    const { candidateId, jobPositionId, scheduledAt, questions } = value;
    const interviewId = uuidv4();

    // Verify candidate and job exist
    const candidate = await new Promise((resolve, reject) => {
      db.get('SELECT * FROM candidates WHERE id = ?', [candidateId], (err, row) => {
        if (err) reject(err);
        else resolve(row);
      });
    });

    const job = await new Promise((resolve, reject) => {
      db.get('SELECT * FROM job_positions WHERE id = ?', [jobPositionId], (err, row) => {
        if (err) reject(err);
        else resolve(row);
      });
    });

    if (!candidate || !job) {
      return res.status(404).json({ error: 'Candidate or job position not found' });
    }

    db.run(
      `INSERT INTO interviews (id, candidate_id, job_position_id, interviewer_id, questions, scheduled_at, status)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [
        interviewId,
        candidateId,
        jobPositionId,
        req.user.userId,
        questions ? JSON.stringify(questions) : null,
        scheduledAt,
        'scheduled'
      ],
      function(err) {
        if (err) {
          console.error('Database error:', err);
          return res.status(500).json({ error: 'Failed to schedule interview' });
        }

        res.status(201).json({
          message: 'Interview scheduled successfully',
          interview: {
            id: interviewId,
            candidateId,
            jobPositionId,
            interviewerId: req.user.userId,
            scheduledAt,
            status: 'scheduled'
          }
        });
      }
    );
  } catch (error) {
    console.error('Schedule interview error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get all interviews
router.get('/', authenticateToken, (req, res) => {
  const page = parseInt(req.query.page) || 1;
  const limit = parseInt(req.query.limit) || 10;
  const offset = (page - 1) * limit;
  const status = req.query.status;

  let query = `
    SELECT i.*, 
           c.name as candidate_name, c.email as candidate_email,
           j.title as job_title,
           u.name as interviewer_name
    FROM interviews i
    JOIN candidates c ON i.candidate_id = c.id
    JOIN job_positions j ON i.job_position_id = j.id
    JOIN users u ON i.interviewer_id = u.id
  `;
  
  let countQuery = 'SELECT COUNT(*) as total FROM interviews i';
  let params = [];

  if (status) {
    query += ' WHERE i.status = ?';
    countQuery += ' WHERE i.status = ?';
    params = [status];
  }

  query += ' ORDER BY i.scheduled_at DESC LIMIT ? OFFSET ?';
  params.push(limit, offset);

  // Get total count
  db.get(countQuery, status ? [status] : [], (err, countResult) => {
    if (err) {
      return res.status(500).json({ error: 'Database error' });
    }

    // Get interviews
    db.all(query, params, (err, interviews) => {
      if (err) {
        return res.status(500).json({ error: 'Database error' });
      }

      // Parse JSON fields
      const processedInterviews = interviews.map(interview => ({
        ...interview,
        questions: interview.questions ? JSON.parse(interview.questions) : null,
        answers: interview.answers ? JSON.parse(interview.answers) : null,
        ai_analysis: interview.ai_analysis ? JSON.parse(interview.ai_analysis) : null
      }));

      res.json({
        interviews: processedInterviews,
        pagination: {
          page,
          limit,
          total: countResult.total,
          pages: Math.ceil(countResult.total / limit)
        }
      });
    });
  });
});

// Get interview by ID
router.get('/:id', authenticateToken, (req, res) => {
  db.get(
    `SELECT i.*, 
           c.name as candidate_name, c.email as candidate_email, c.phone as candidate_phone,
           j.title as job_title, j.description as job_description,
           u.name as interviewer_name
     FROM interviews i
     JOIN candidates c ON i.candidate_id = c.id
     JOIN job_positions j ON i.job_position_id = j.id
     JOIN users u ON i.interviewer_id = u.id
     WHERE i.id = ?`,
    [req.params.id],
    (err, interview) => {
      if (err) {
        return res.status(500).json({ error: 'Database error' });
      }

      if (!interview) {
        return res.status(404).json({ error: 'Interview not found' });
      }

      // Parse JSON fields
      interview.questions = interview.questions ? JSON.parse(interview.questions) : null;
      interview.answers = interview.answers ? JSON.parse(interview.answers) : null;
      interview.ai_analysis = interview.ai_analysis ? JSON.parse(interview.ai_analysis) : null;

      res.json({ interview });
    }
  );
});

router.post('/from-assessment', authenticateToken, async (req, res) => {
  try {
    const { candidateId, jobId, scheduledAt } = req.body;

    // Validate required fields
    if (!candidateId || !jobId || !scheduledAt) {
      return res.status(400).json({ 
        error: 'Candidate ID, Job ID, and scheduled date/time are required' 
      });
    }

    // Verify candidate exists
    const candidate = await new Promise((resolve, reject) => {
      db.get('SELECT * FROM candidates WHERE id = ?', [candidateId], (err, row) => {
        if (err) reject(err);
        else resolve(row);
      });
    });

    if (!candidate) {
      return res.status(404).json({ error: 'Candidate not found' });
    }

    // Verify job exists
    const job = await new Promise((resolve, reject) => {
      db.get('SELECT * FROM job_positions WHERE id = ?', [jobId], (err, row) => {
        if (err) reject(err);
        else resolve(row);
      });
    });

    if (!job) {
      return res.status(404).json({ error: 'Job position not found' });
    }

    // Create interview record
    const interviewId = uuidv4();
    
    await new Promise((resolve, reject) => {
      db.run(
        `INSERT INTO interviews (id, candidate_id, job_position_id, interviewer_id, scheduled_at, status)
         VALUES (?, ?, ?, ?, ?, ?)`,
        [
          interviewId,
          candidateId,
          jobId,
          req.user.userId,
          new Date(scheduledAt).toISOString(),
          'scheduled'
        ],
        function(err) {
          if (err) reject(err);
          else resolve();
        }
      );
    });

    console.log(`✅ Interview scheduled from assessment for candidate ${candidateId}, interview ID: ${interviewId}`);

    res.status(201).json({
      message: 'Interview scheduled successfully from assessment',
      interview: {
        id: interviewId,
        candidateId,
        candidateName: candidate.name,
        jobId,
        jobTitle: job.title,
        interviewerId: req.user.userId,
        scheduledAt: new Date(scheduledAt).toISOString(),
        status: 'scheduled'
      }
    });

  } catch (error) {
    console.error('Schedule interview from assessment error:', error);
    res.status(500).json({ error: error.message || 'Failed to schedule interview' });
  }
});

// Start interview (update status to in_progress)
router.post('/:id/start', authenticateToken, (req, res) => {
  db.run(
    'UPDATE interviews SET status = ? WHERE id = ? AND interviewer_id = ?',
    ['in_progress', req.params.id, req.user.userId],
    function(err) {
      if (err) {
        return res.status(500).json({ error: 'Failed to start interview' });
      }

      if (this.changes === 0) {
        return res.status(404).json({ error: 'Interview not found or unauthorized' });
      }

      res.json({ message: 'Interview started successfully' });
    }
  );
});

// Submit answer and get AI analysis
router.post('/:id/answer', authenticateToken, async (req, res) => {
  try {
    const { error, value } = answerSchema.validate(req.body);
    if (error) {
      return res.status(400).json({ error: error.details[0].message });
    }

    const { questionId, question, answer } = value;
    const interviewId = req.params.id;

    // Get interview and job context
    const interview = await new Promise((resolve, reject) => {
      db.get(
        `SELECT i.*, j.title, j.description 
         FROM interviews i 
         JOIN job_positions j ON i.job_position_id = j.id 
         WHERE i.id = ?`,
        [interviewId],
        (err, row) => {
          if (err) reject(err);
          else resolve(row);
        }
      );
    });

    if (!interview) {
      return res.status(404).json({ error: 'Interview not found' });
    }

    // Get AI analysis of the answer
    const jobContext = `${interview.title}: ${interview.description}`;
    const analysis = await groqService.analyzeInterviewResponse(question, answer, jobContext);

    // Update interview with new answer
    let currentAnswers = interview.answers ? JSON.parse(interview.answers) : [];
    currentAnswers.push({
      id: questionId,
      question,
      answer,
      analysis,
      timestamp: new Date().toISOString()
    });

    db.run(
      'UPDATE interviews SET answers = ? WHERE id = ?',
      [JSON.stringify(currentAnswers), interviewId],
      function(err) {
        if (err) {
          console.error('Database error:', err);
          return res.status(500).json({ error: 'Failed to save answer' });
        }

        res.json({
          message: 'Answer submitted and analyzed successfully',
          analysis
        });
      }
    );
  } catch (error) {
    console.error('Submit answer error:', error);
    res.status(500).json({ error: error.message || 'Failed to analyze answer' });
  }
});

// Complete interview
router.post('/:id/complete', authenticateToken, async (req, res) => {
  try {
    const interviewId = req.params.id;
    const { overallFeedback, finalScore } = req.body;

    // Get interview data
    const interview = await new Promise((resolve, reject) => {
      db.get('SELECT * FROM interviews WHERE id = ?', [interviewId], (err, row) => {
        if (err) reject(err);
        else resolve(row);
      });
    });

    if (!interview) {
      return res.status(404).json({ error: 'Interview not found' });
    }

    // Calculate overall analysis
    const answers = interview.answers ? JSON.parse(interview.answers) : [];
    let totalScore = 0;
    let analysisCount = 0;

    answers.forEach(answer => {
      if (answer.analysis && answer.analysis.score) {
        totalScore += answer.analysis.score;
        analysisCount++;
      }
    });

    const averageScore = analysisCount > 0 ? Math.round((totalScore / analysisCount) * 10) : finalScore || 0;

    const overallAnalysis = {
      averageScore,
      finalScore: finalScore || averageScore,
      totalQuestions: answers.length,
      overallFeedback: overallFeedback || 'No additional feedback provided',
      completedAt: new Date().toISOString(),
      answersSummary: answers.map(a => ({
        question: a.question,
        score: a.analysis?.score || 0,
        strengths: a.analysis?.strengths || [],
        improvements: a.analysis?.improvements || []
      }))
    };

    db.run(
      'UPDATE interviews SET status = ?, ai_analysis = ?, score = ?, completed_at = CURRENT_TIMESTAMP WHERE id = ?',
      ['completed', JSON.stringify(overallAnalysis), averageScore, interviewId],
      function(err) {
        if (err) {
          console.error('Database error:', err);
          return res.status(500).json({ error: 'Failed to complete interview' });
        }

        res.json({
          message: 'Interview completed successfully',
          analysis: overallAnalysis
        });
      }
    );
  } catch (error) {
    console.error('Complete interview error:', error);
    res.status(500).json({ error: 'Failed to complete interview' });
  }
});

// Update interview
router.put('/:id', authenticateToken, async (req, res) => {
  try {
    const interviewId = req.params.id;
    const { scheduledAt, questions, status } = req.body;

    let updateFields = [];
    let updateValues = [];

    if (scheduledAt) {
      updateFields.push('scheduled_at = ?');
      updateValues.push(scheduledAt);
    }

    if (questions) {
      updateFields.push('questions = ?');
      updateValues.push(JSON.stringify(questions));
    }

    if (status) {
      updateFields.push('status = ?');
      updateValues.push(status);
    }

    if (updateFields.length === 0) {
      return res.status(400).json({ error: 'No fields to update' });
    }

    updateValues.push(interviewId);
    updateValues.push(req.user.userId);

    db.run(
      `UPDATE interviews SET ${updateFields.join(', ')} WHERE id = ? AND interviewer_id = ?`,
      updateValues,
      function(err) {
        if (err) {
          return res.status(500).json({ error: 'Failed to update interview' });
        }

        if (this.changes === 0) {
          return res.status(404).json({ error: 'Interview not found or unauthorized' });
        }

        res.json({ message: 'Interview updated successfully' });
      }
    );
  } catch (error) {
    console.error('Update interview error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Delete interview
router.delete('/:id', authenticateToken, (req, res) => {
  db.run(
    'DELETE FROM interviews WHERE id = ? AND interviewer_id = ?',
    [req.params.id, req.user.userId],
    function(err) {
      if (err) {
        return res.status(500).json({ error: 'Failed to delete interview' });
      }

      if (this.changes === 0) {
        return res.status(404).json({ error: 'Interview not found or unauthorized' });
      }

      res.json({ message: 'Interview deleted successfully' });
    }
  );
});
module.exports = router;